plugins/pulp_docker/plugins/distributors/publish_steps.py

```diff
@@ -220,3 +223,154 @@ def process_main(self):
+        self.parent.symlink_list.append(symlink)
+        self._tag_names.add(item.name)
+
+    def finalize(self):
+        """
+        Write the Tag list file so that clients can retrieve the list of available Tags.
+        """
+        tags_path = os.path.join(self.parent.get_working_dir(), '.relative', 'tags')
+        misc.mkdir(tags_path)
+        with open(os.path.join(tags_path, 'list'), 'w') as list_file:
+            tag_data = {
+                'name': self.repo_registry_id,
+                'tags': list(self._tag_names)}
+            list_file.write(json.dumps(tag_data))
+        # We don't need the tag names anymore
+        del self._tag_names
+
+
+class DockerRsyncPublisher(Publisher):
+
+    REPO_CONTENT_TYPES = [constants.IMAGE_TYPE_ID, constants.BLOB_TYPE_ID,
+                          constants.MANIFEST_TYPE_ID]
+
+    REPO_CONTENT_MODELS = [models.Blob, models.Manifest, models.Image]
+
+    def _get_predistributor(self):
+        """
+        Returns the distributor object representing the predistributor. A predistributor is the
+        distributor that is used to publish a repository prior to publishing with the rsync
+        distributor.
+
+        :return: predistributor that was configured in rsync distributor's config
+        :rtype: Distributor
+        """
+        return None
+
+    def _get_postdistributor(self):
+        """
+        Returns the distributor object representing the postdistributor. A postdistirbutor is the
+        distributor used to publish the repository after an rsync publish occurs.
+
+        :return: postdistributor that was configured in rsync distributor's config
+        :rtype: Distributor
+        """
+        postdistributor_id = self.get_config().flatten().get("postdistributor_id", None)
+        return Distributor.objects.get_or_404(repo_id=self.repo.id,
+                                              distributor_id=postdistributor_id)
+
+    def _add_necesary_steps(self, date_filter=None, relative_symlinks=True, config=None):
+        """
+        This method adds all the steps that are needed to accomplish an RPM rsync publish. This
+        includes:
+
+        Unit Query Step - selects units associated with the repo based on the date_filter and
+                          creates relative symlinks
+        Tag Generation Step  - creates relative symlinks for for all tags and writes out list file
+        Rsync Step (content units) - rsyncs content units from /var/lib/pulp/content to remote
+                                     server
+        Rsync Step (symlinks) - rsyncs symlinks from working directory to remote server
+
+        :param date_filter:  Q object with start and/or end dates, or None if start and end dates
+                             are not provided
+        :type date_filter:  mongoengine.Q or types.NoneType
+        :param config: distributor configuration
+        :type config: pulp.plugins.config.PluginCallConfiguration
+        :return: None
```

Comment:
Consider separating the input and output arguments with a line of whitespace for readability.